README.md
# StackRecon
Python to collect web daemon, eCommerce application configuration, and supporting MySQL and no-sql daemon information.

# Information pooled as of June 2, 2016
Information is dynamically discovered, instead of using static names or paths. It is saved to a json file, so it is easy to save the raw information for cross reference.
ecommStackStatus should be run on the web nodes, and it will attempt to connect to other services over the network as needed, such as MySQL, memcache and redis.

nginx & apache
URL, document root, configuration file, log files, and listening ports

php-fpm
Pool information for php-fpm.

Memory profile for nginx, apache and php-fpm. This is most useful for the php process, either mod_php or php-fpm.

Magento queries for each document root
Database configuration, cache configuration for session, objects and full page cache, and whether the cache is enabled in the Magento configuration portal.

For each Magento document root:
MySQL host, port and credential information
Memcache host and port information
Redis host and port information
Magento's MySQL table for cache configuration (enabled or disabled)

Additional memcache and redis information



changelog.txt
TO-DO List
Add checks for eth0 or bond0 IP vs the database and cache IPs
Add per server check for innodb data+index vs innodb buffer size
MySQL max connections check
mysql-recon

2016-09-28
Checks for cron job, should be cron.sh NOT cron.php
Magento 2 support is complete
MySQL is now queried with mysql-connector instead of the mysql binary on the client machines

2016-09-08
The project name changed from ecommStackStatus to stackrecon to better reflect its capabilities. 
The classic ecommStackStatus will be ecomm-recon and Apache, Nginx and PHP capabilities will be available as web-recon.
The library is now named stackreconlib. This should make it easier to separate it and create a pip library in the future.

Magento 2 support is coming along. I added version parsing today.

* TODO things to add
php shell/indexer.php --status

We could get information similar to MySQL Buddy and display it, to name a few:
long_query_time
query_cache_size
join_buffer_size
table_open_cache
innodb_buffer_pool_size
innodb_buffer_pool_instances
innodb_log_buffer_size
query_cache_limit
* Magento report numbers for reports in the last 24-48 hours with date and time
* name json file by hostname and date+time
* I would like to load all xml in app/etc/ and overwrite values with local.xml so the config is complete
* Varnish detection and cache health
# ps -ef|grep [v]arnish
root     11893     1  0 Nov25 ?        00:05:35 /usr/sbin/varnishd -P /var/run/varnish.pid -a :80 -f /etc/varnish/default.vcl -T 192.168.100.168:6082 -t 120 -w 50,1000,120 -u varnish -g varnish -p cli_buffer=16384 -S /etc/varnish/secret -s malloc,10G
varnish  11894 11893  2 Nov25 ?        02:45:04 /usr/sbin/varnishd -P /var/run/varnish.pid -a :80 -f /etc/varnish/default.vcl -T 192.168.100.168:6082 -t 120 -w 50,1000,120 -u varnish -g varnish -p cli_buffer=16384 -S /etc/varnish/secret -s malloc,10G
* Add mysql branch to globalconfig, and parse "show variables;"
proposed structure:
mysql: {
    HOSTNAME: {
        port: "", # Do I need this? It is nearly always 3306
        username: "",
        password: "",
        variables: {
            `show variables` # parsed to key:value pairs
        }
    }
}
* MySQL max_connections, max_used_connections
* MySQL query cache, example values: query_cache_type=1, query_cache_size=256M, query_cache_limit=16M
* Check Magento for the Shoplift SUPEE-5344 vulnerability
find /var/www -wholename '*/app/code/core/Mage/Core/Controller/Request/Http.php' | xargs grep -L _internallyForwarded
If it returns results, assuming Magento is in /var/www, it is vulnerable.
-L Suppress normal output; instead print the name of each input file from which no output would normally have been printed.  The scanning will stop on the first match.
Check doc_root/app/code/core/Mage/Core/Controller/Request/Http.php
If it doesn't have _internallyForwarded it is probably vulnerable to shoplift
* Check Magento for SUPEE-7405
* check php opcache
i.e.
Re-enabled PHP opcache in /etc/php.d/10-opcache.ini:
opcache.enable=1
Changed the "0" to a "1" on that line.
Stop nginx, restart php-fpm, start nginx.
* check mysql
* magento_root/shell/indexer.php --status
i.e.
2560M
2024M
Category Flat Data:                 Pending
Product Flat Data:                  Pending
Stock Status:                       Pending
Catalog product price:              Pending
Category URL Rewrites:              Pending
Product URL Rewrites:               Pending
URL Redirects:                      Pending
Catalog Category/Product Index:     Pending
Catalog Search Index:               Pending
Default Values (MANAdev):           Pending
Dynamic Categories:                 Running
Tag Aggregation Data:               Pending
SEO Schemas (MANAdev):              Pending
Product Attributes:                 Pending
SEO URL Rewrites (MANAdev):         Pending

DONE
* also need to check, if session cache is using redis - DONE
app/etc/modules/Cm_RedisSessions.xml
value of <active> to true
* add hostname in globalconfig
* Parse this session_cache syntax for redis
Session Cache engine: unknown
Session Cache: redis
session_save: redis
session_save_path: tcp://192.168.100.200:6379?weight=2&timeout=2.5
From local.xml:
        <session_save><![CDATA[redis]]></session_save>
        <session_save_path><![CDATA[tcp://192.168.100.200:6379?weight=2&timeout=2.5]]></session_save_path>


processlist.py
#!/usr/bin/env python2

import re
import os

#changed the name, and have a duplicate, but I spent too long hunting bugs related to the wrong name.
def daemon_exe(match_exe):
    """
    var_filter = "text to search with"
    using this as the filter will find an executable by name whether it was call by absolute path or bare
    "^(\S*/bash|bash)"
    """
    daemons = {}
    pids = [pid for pid in os.listdir('/proc') if pid.isdigit()]
    
    for pid in pids:
        try:
            pscmd = open(os.path.join('/proc', pid, 'cmdline'), 'rb').read().replace("\000"," ").rstrip()
            psexe = os.path.realpath(os.path.join('/proc', pid, 'exe'))
        except (IOError,OSError): # proc has already terminated, you may not be root
            continue
        if psexe:
            for daemon_name in match_exe:
                if os.path.basename(psexe) == daemon_name:
                    if not "daemon_name" in daemons:
                        daemons[daemon_name] = { "exe" : [], "cmd" : [] }
                    daemons[daemon_name]["exe"] += [os.path.basename(psexe)]
                    daemons[daemon_name]["cmd"] += [pscmd]
    return(daemons)

# works
#pslist = processlist("^(\S*/bash|bash)")
#for key in pslist.keys():
#    print "%6d %s" % (int(key),pslist[key])

# also works
#for key, value in processlist("^(\S*/bash|bash)").iteritems():
#    print "%6d %s" % (int(key),value)
    
# works, no filter
#for key, value in processlist().iteritems():
#    print "%6d %s" % (int(key),value)

import psutil
import os
var_filter = "http"

def daemon_exe(match_exe):
    daemons = {}
    
    pidlist = psutil.get_pid_list()
    for pid in pidlist:
        p = psutil.Process(pid)
        try:
            if p.exe:
                #match_exe = ["httpd", "apache2", "nginx", "bash"]
                for daemon_name in match_exe:
                    if os.path.basename(p.exe) == daemon_name:
                        if not "daemon_name" in daemons:
                            daemons[daemon_name] = { "exe" : [] }
                        daemons[daemon_name]["exe"] += [p.exe]
                        #print p.exe
        except:
            #print "You should run this as root"
            pass
    return daemons

daemons = daemon_exe(["httpd", "apache2", "nginx", "bash"])
daemons = processlist(["httpd", "apache2", "nginx", "bash"])
#for key,value in daemons.iteritems():
#    print "%s %r" % (key,value)
if "apache2" in daemons:
    print "apache2 on Debian/Ubuntu %s" % daemons["apache2"]["exe"][0]
if "httpd" in daemons:
    print "httpd on Red Hat/CentOS %s" % daemons["httpd"]["exe"][0]
    
    
    
"""
       /proc/[pid]/stat
              Status information about the process.  This is used by ps(1).
              It is defined in the kernel source file fs/proc/array.c.

              The fields, in order, with their proper scanf(3) format
              specifiers, are:

              (1) pid  %d
                        The process ID.

              (2) comm  %s
                        The filename of the executable, in parentheses.
                        This is visible whether or not the executable is
                        swapped out.

              (3) state  %c
                        One of the following characters, indicating process
                        state:

                        R  Running

                        S  Sleeping in an interruptible wait

                        D  Waiting in uninterruptible disk sleep

                        Z  Zombie

                        T  Stopped (on a signal) or (before Linux 2.6.33)
                           trace stopped

                        t  Tracing stop (Linux 2.6.33 onward)

                        W  Paging (only before Linux 2.6.0)

                        X  Dead (from Linux 2.6.0 onward)

                        x  Dead (Linux 2.6.33 to 3.13 only)

                        K  Wakekill (Linux 2.6.33 to 3.13 only)

                        W  Waking (Linux 2.6.33 to 3.13 only)

                        P  Parked (Linux 3.9 to 3.13 only)

              (4) ppid  %d
                        The PID of the parent of this process.

              (5) pgrp  %d
                        The process group ID of the process.

              (6) session  %d
                        The session ID of the process.

              (7) tty_nr  %d
                        The controlling terminal of the process.  (The minor
                        device number is contained in the combination of
                        bits 31 to 20 and 7 to 0; the major device number is
                        in bits 15 to 8.)

              (8) tpgid  %d
                        The ID of the foreground process group of the
                        controlling terminal of the process.

              (9) flags  %u
                        The kernel flags word of the process.  For bit
                        meanings, see the PF_* defines in the Linux kernel
                        source file include/linux/sched.h.  Details depend
                        on the kernel version.

                        The format for this field was %lu before Linux 2.6.

              (10) minflt  %lu
                        The number of minor faults the process has made
                        which have not required loading a memory page from
                        disk.

              (11) cminflt  %lu
                        The number of minor faults that the process's
                        waited-for children have made.

              (12) majflt  %lu
                        The number of major faults the process has made
                        which have required loading a memory page from disk.

              (13) cmajflt  %lu
                        The number of major faults that the process's
                        waited-for children have made.

              (14) utime  %lu
                        Amount of time that this process has been scheduled
                        in user mode, measured in clock ticks (divide by
                        sysconf(_SC_CLK_TCK)).  This includes guest time,
                        guest_time (time spent running a virtual CPU, see
                        below), so that applications that are not aware of
                        the guest time field do not lose that time from
                        their calculations.

              (15) stime  %lu
                        Amount of time that this process has been scheduled
                        in kernel mode, measured in clock ticks (divide by
                        sysconf(_SC_CLK_TCK)).

              (16) cutime  %ld
                        Amount of time that this process's waited-for
                        children have been scheduled in user mode, measured
                        in clock ticks (divide by sysconf(_SC_CLK_TCK)).
                        (See also times(2).)  This includes guest time,
                        cguest_time (time spent running a virtual CPU, see
                        below).

              (17) cstime  %ld
                        Amount of time that this process's waited-for
                        children have been scheduled in kernel mode,
                        measured in clock ticks (divide by
                        sysconf(_SC_CLK_TCK)).

              (18) priority  %ld
                        (Explanation for Linux 2.6) For processes running a
                        real-time scheduling policy (policy below; see
                        sched_setscheduler(2)), this is the negated
                        scheduling priority, minus one; that is, a number in
                        the range -2 to -100, corresponding to real-time
                        priorities 1 to 99.  For processes running under a
                        non-real-time scheduling policy, this is the raw
                        nice value (setpriority(2)) as represented in the
                        kernel.  The kernel stores nice values as numbers in
                        the range 0 (high) to 39 (low), corresponding to the
                        user-visible nice range of -20 to 19.

                        Before Linux 2.6, this was a scaled value based on
                        the scheduler weighting given to this process.

              (19) nice  %ld
                        The nice value (see setpriority(2)), a value in the
                        range 19 (low priority) to -20 (high priority).

              (20) num_threads  %ld
                        Number of threads in this process (since Linux 2.6).
                        Before kernel 2.6, this field was hard coded to 0 as
                        a placeholder for an earlier removed field.

              (21) itrealvalue  %ld
                        The time in jiffies before the next SIGALRM is sent
                        to the process due to an interval timer.  Since
                        kernel 2.6.17, this field is no longer maintained,
                        and is hard coded as 0.

              (22) starttime  %llu
                        The time the process started after system boot.  In
                        kernels before Linux 2.6, this value was expressed
                        in jiffies.  Since Linux 2.6, the value is expressed
                        in clock ticks (divide by sysconf(_SC_CLK_TCK)).

                        The format for this field was %lu before Linux 2.6.

              (23) vsize  %lu
                        Virtual memory size in bytes.

              (24) rss  %ld
                        Resident Set Size: number of pages the process has
                        in real memory.  This is just the pages which count
                        toward text, data, or stack space.  This does not
                        include pages which have not been demand-loaded in,
                        or which are swapped out.

              (25) rsslim  %lu
                        Current soft limit in bytes on the rss of the
                        process; see the description of RLIMIT_RSS in
                        getrlimit(2).

              (26) startcode  %lu
                        The address above which program text can run.

              (27) endcode  %lu
                        The address below which program text can run.

              (28) startstack  %lu
                        The address of the start (i.e., bottom) of the
                        stack.

              (29) kstkesp  %lu
                        The current value of ESP (stack pointer), as found
                        in the kernel stack page for the process.

              (30) kstkeip  %lu
                        The current EIP (instruction pointer).

              (31) signal  %lu
                        The bitmap of pending signals, displayed as a
                        decimal number.  Obsolete, because it does not
                        provide information on real-time signals; use
                        /proc/[pid]/status instead.

              (32) blocked  %lu
                        The bitmap of blocked signals, displayed as a
                        decimal number.  Obsolete, because it does not
                        provide information on real-time signals; use
                        /proc/[pid]/status instead.

              (33) sigignore  %lu
                        The bitmap of ignored signals, displayed as a
                        decimal number.  Obsolete, because it does not
                        provide information on real-time signals; use
                        /proc/[pid]/status instead.

              (34) sigcatch  %lu
                        The bitmap of caught signals, displayed as a decimal
                        number.  Obsolete, because it does not provide
                        information on real-time signals; use
                        /proc/[pid]/status instead.

              (35) wchan  %lu
                        This is the "channel" in which the process is
                        waiting.  It is the address of a location in the
                        kernel where the process is sleeping.  The
                        corresponding symbolic name can be found in
                        /proc/[pid]/wchan.

              (36) nswap  %lu
                        Number of pages swapped (not maintained).

              (37) cnswap  %lu
                        Cumulative nswap for child processes (not
                        maintained).

              (38) exit_signal  %d  (since Linux 2.1.22)
                        Signal to be sent to parent when we die.

              (39) processor  %d  (since Linux 2.2.8)
                        CPU number last executed on.

              (40) rt_priority  %u  (since Linux 2.5.19)
                        Real-time scheduling priority, a number in the range
                        1 to 99 for processes scheduled under a real-time
                        policy, or 0, for non-real-time processes (see
                        sched_setscheduler(2)).

              (41) policy  %u  (since Linux 2.5.19)
                        Scheduling policy (see sched_setscheduler(2)).
                        Decode using the SCHED_* constants in linux/sched.h.

                        The format for this field was %lu before Linux
                        2.6.22.

              (42) delayacct_blkio_ticks  %llu  (since Linux 2.6.18)
                        Aggregated block I/O delays, measured in clock ticks
                        (centiseconds).

              (43) guest_time  %lu  (since Linux 2.6.24)
                        Guest time of the process (time spent running a
                        virtual CPU for a guest operating system), measured
                        in clock ticks (divide by sysconf(_SC_CLK_TCK)).

              (44) cguest_time  %ld  (since Linux 2.6.24)
                        Guest time of the process's children, measured in
                        clock ticks (divide by sysconf(_SC_CLK_TCK)).

              (45) start_data  %lu  (since Linux 3.3)
                        Address above which program initialized and
                        uninitialized (BSS) data are placed.

              (46) end_data  %lu  (since Linux 3.3)
                        Address below which program initialized and
                        uninitialized (BSS) data are placed.

              (47) start_brk  %lu  (since Linux 3.3)
                        Address above which program heap can be expanded
                        with brk(2).

              (48) arg_start  %lu  (since Linux 3.5)
                        Address above which program command-line arguments
                        (argv) are placed.

              (49) arg_end  %lu  (since Linux 3.5)
                        Address below program command-line arguments (argv)
                        are placed.

              (50) env_start  %lu  (since Linux 3.5)
                        Address above which program environment is placed.

              (51) env_end  %lu  (since Linux 3.5)
                        Address below which program environment is placed.

              (52) exit_code  %d  (since Linux 3.5)
                        The thread's exit status in the form reported by
                        waitpid(2).
"""